### PR TITLE
FIFO cleanup and documentation

### PR DESCRIPTION
**Pull request by pkendall64**
# Refactoring
This PR does three things...

1. Refactoring the FIFO (of which we had 2 implemenations) into a single template class.
2. Introduces `lock`/`unlock` semantics on the FIFO for more coarse-grained atomic operations on the FIFO. This allows for better compiler optimisation and safety.
3. Document the API (I'm on a crusade!)

As a side-effect, it also reduces flash/RAM usage on the poor STM devices too!

## Ghost ATTO Receiver
before...
```
RAM:   [=====     ]  46.1% (used 7560 bytes from 16384 bytes)
Flash: [=======   ]  72.9% (used 47752 bytes from 65536 bytes)
```
after...
```
RAM:   [====      ]  44.5% (used 7296 bytes from 16384 bytes)
Flash: [=======   ]  72.4% (used 47480 bytes from 65536 bytes)
```
272 bytes for flash saved and 264 bytes of RAM!

## R9M Transmitter
before...
```
RAM:   [===       ]  31.3% (used 6408 bytes from 20480 bytes)
Flash: [=======   ]  72.2% (used 47320 bytes from 65536 bytes)
```
after...
```
RAM:   [===       ]  31.4% (used 6428 bytes from 20480 bytes)
Flash: [=======   ]  72.1% (used 47248 bytes from 65536 bytes)
```
72 bytes saved for flash and 20 bytes more of RAM

### CODE REVIEW

**Comment by JyeSmith** [comment]
> Document the API (I'm on a crusade!)

❤️

File: src/lib/CRSF/CRSF.cpp [hunk 1]
```diff
@@ -32,7 +30,7 @@ HardwareSerial CRSF::Port = Serial;
 GENERIC_CRC8 crsf_crc(CRSF_CRC_POLY);
 
 /// Out FIFO to buffer messages///
-static FIFO SerialOutFIFO;
+static FIFO<256> SerialOutFIFO;
```
Comment (wvarty) [inline src/lib/CRSF/CRSF.cpp:33]: 256 seems to be used in >1 places in this file.... should it be turned into a `#define`?
Also... DEATH TO MAGIC NUMBERS!

File: src/src/rx-serial/SerialIO.h [hunk 2]
```diff
@@ -24,7 +24,7 @@ class SerialIO {
 protected:
     Stream *_outputPort;
     Stream *_inputPort;
-    FIFO _fifo;
+    FIFO<256> _fifo;
```
Comment (wvarty) [inline src/src/rx-serial/SerialIO.h:27]: More magic numbers here

File: src/lib/CRSF/CRSF.cpp (hunk 1 above)
Comment (pkendall64) [reply src/lib/CRSF/CRSF.cpp]: Added constants for magic numbers 💀

File: src/src/rx-serial/SerialIO.h (hunk 2 above)
Comment (pkendall64) [reply src/src/rx-serial/SerialIO.h]: Added constants for magic numbers 💀

**Comment by pkendall64** [comment]
Airport tested with Ranger & DIY 8285 RX @ 9600 baud F500
```
Bytes/s =  808 Detected drops =  1 expected =  9 received =  9 sent =  27 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  10 received =  10 sent =  28 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  11 received =  11 sent =  29 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  12 received =  12 sent =  30 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  13 received =  13 sent =  31 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  14 received =  14 sent =  32 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  15 received =  15 sent =  33 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  16 received =  16 sent =  34 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  17 received =  17 sent =  35 bufSizeByte =  b''
Bytes/s =  808 Detected drops =  1 expected =  18 received =  18 sent =  36 bufSizeByte =  b''
```

**Review by wvarty — APPROVED** [review]

**Review by JyeSmith — APPROVED** [review]

**Pull request merged**